### ISSUE TITLE
[MAS] Occasional JVM crashes during SpeechRecognizer dispose when using MAS

### ISSUE THREAD

**Issue by RomanValov** [Azure-Samples/cognitive-services-speech-sdk#2125]
Hi there!

This is somehow related to #2124 but in contrast to other issues this one reproduced when using continuous recognition mode during clean up:

```java
package com.speech.demo;

import com.microsoft.cognitiveservices.speech.audio.*;
import com.microsoft.cognitiveservices.speech.*;


public class Main {

    private static boolean pass = false;

    private static SpeechConfig initSpeechConfig(String speechSubscriptionKey, String serviceRegion) {
        SpeechConfig speechConfig = SpeechConfig.fromSubscription(speechSubscriptionKey, serviceRegion);
        speechConfig.setProperty(PropertyId.Speech_LogFilename, "/tmp/speech.log");
        return speechConfig;
    }

    public static void feed(PushAudioInputStream input) {
        byte[] buf = new byte[640];
        while (true) {
            try {
                Thread.sleep(20);
                if (pass) {
                    input.write(buf);
                }
            } catch (Exception ex) {
            }
        }
    }

    public static void main(String[] args) {

        String speechSubscriptionKey = System.getenv("SPEECH_KEY");
        String serviceRegion = System.getenv("SPEECH_REGION");
        boolean useMas = Boolean.parseBoolean(System.getenv("SPEECH_USE_MAS"));

        AudioStreamFormat format = AudioStreamFormat.getWaveFormat(16000, (short)16, (short)1, AudioStreamWaveFormat.PCM);

        try (
            final PushAudioInputStream input = PushAudioInputStream.create(format);
            final AudioProcessingOptions audioOpts = useMas ? AudioProcessingOptions.create(
                AudioProcessingConstants.AUDIO_INPUT_PROCESSING_ENABLE_DEFAULT,
                PresetMicrophoneArrayGeometry.Mono,
                SpeakerReferenceChannel.None) : null;
            final AudioConfig audioConfig = useMas ? AudioConfig.fromStreamInput(input, audioOpts) :
                                                    AudioConfig.fromStreamInput(input);
            final SpeechConfig speechConfig = initSpeechConfig(speechSubscriptionKey, serviceRegion);
            final SpeechRecognizer reco = new SpeechRecognizer(speechConfig, audioConfig);
        ) {
            System.out.printf("MAS: %s %s\n", useMas, audioOpts);

            reco.recognized.addEventListener((s, e) -> {
                SpeechRecognitionResult result = e.getResult();
                if (result.getReason() == ResultReason.RecognizedSpeech) {
                    System.out.println("RECOGNIZED: " + result.getText());
                }
                else if (result.getReason() == ResultReason.NoMatch) {
                    NoMatchDetails noMatch = NoMatchDetails.fromResult(result);
                    System.out.println("NOMATCH: Reason=" + noMatch.getReason());
                }
            });

            reco.canceled.addEventListener((s, e) -> {
                System.out.println("CANCELED: Reason=" + e.getReason());

                if (e.getReason() == CancellationReason.Error) {
                    System.out.println("CANCELED: ErrorCode=" + e.getErrorCode());
                    System.out.println("CANCELED: ErrorDetails=" + e.getErrorDetails());
                }
            });

            reco.sessionStarted.addEventListener((s, e) -> {
                System.out.println("Session started");
                pass = true;
            });

            reco.sessionStopped.addEventListener((s, e) -> {
                pass = false;
                System.out.println("Session stopped");
            });

            Thread feedThread = new Thread(() -> feed(input));
            feedThread.setDaemon(true);
            feedThread.start();

            Thread.sleep(2500);

            reco.startContinuousRecognitionAsync().get();

            Thread.sleep(2500);

            reco.stopContinuousRecognitionAsync().get();
        } catch (Exception ex) {
            throw new RuntimeException(ex);
        }

        format.close();
    }

}
```

I'm running this code in a Docker container with Ubuntu Linux 20.04 and OpenJDK 11.
It's supposed it doesn't matter if specific input audio stream to be used. Sample code generates zeroed by stream.
The issue is not 100% reproducible but when it is it crashes all the JVM environment, not the app on it's own.
So in order to reproduce the crash I run the program in a loop like:

```
while :; do docker run -e SPEECH_USE_MAS=... ... || break; done
```

**Expected behavior**

When not using MAS everything works fine:

```
MAS: false null
Session started
NOMATCH: Reason=InitialSilenceTimeout
Session stopped
MAS: false null
Session started
NOMATCH: Reason=InitialSilenceTimeout
Session stopped
MAS: false null
Session started
NOMATCH: Reason=InitialSilenceTimeout
Session stopped
MAS: false null
Session started
NOMATCH: Reason=InitialSilenceTimeout
Session stopped
...
```

Log: 
[speech.success.log](https://github.com/Azure-Samples/cognitive-services-speech-sdk/files/13250902/speech.success.log)

**Actual behavior**

But when MAS is involved (`SPEECH_USE_MAS=true`) once in a while the program causes JVM to crash:

```
MAS: true com.microsoft.cognitiveservices.speech.audio.AudioProcessingOptions@357246de
Session started
NOMATCH: Reason=InitialSilenceTimeout
Session stopped
MAS: true com.microsoft.cognitiveservices.speech.audio.AudioProcessingOptions@357246de
Session started
NOMATCH: Reason=InitialSilenceTimeout
Session stopped
MAS: true com.microsoft.cognitiveservices.speech.audio.AudioProcessingOptions@357246de
Session started
NOMATCH: Reason=InitialSilenceTimeout
Session stopped
MAS: true com.microsoft.cognitiveservices.speech.audio.AudioProcessingOptions@357246de
Session started
NOMATCH: Reason=InitialSilenceTimeout
Session stopped
terminate called without an active exception
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fa90ed53941, pid=1, tid=7
#
# JRE version: OpenJDK Runtime Environment (11.0.20.1+1) (build 11.0.20.1+1-post-Ubuntu-0ubuntu120.04)
# Java VM: OpenJDK 64-Bit Server VM (11.0.20.1+1-post-Ubuntu-0ubuntu120.04, mixed mode, sharing, tiered, compressed oops, serial gc, linux-amd64)
# Problematic frame:
# C  [libc.so.6+0x22941]  abort+0x213
#
# Core dump will be written. Default location: Core dumps may be processed with "/usr/share/apport/apport -p%p -s%s -c%c -d%d -P%P -u%u -g%g -- %E" (or dumping to /app/core.1)
#
# An error report file with more information is saved as:
# /tmp/hs_err_pid1.log
#
# If you would like to submit a bug report, please visit:
#   https://bugs.launchpad.net/ubuntu/+source/openjdk-lts
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

Log: 
[speech.failure.log](https://github.com/Azure-Samples/cognitive-services-speech-sdk/files/13250920/speech.failure.log)

**Additional details**

The setup of MAS is pretty dummy in the sample merely to reproduce the issue. I've got same results if using `SpeakerReferenceChannel.LastChannel`. It seems that whether crash is reproduced or not is a subject to fine tuning sleep delays.

**Version of the Cognitive Services Speech SDK**

1.31.0

**Platform, Operating System, and Programming Language**

 - OS: Ubuntu Linux 20.04 (docker)
 - Hardware - x64
 - Programming language: Java OpenJDK 11

**Comment by pankopon** [comment]
@jychoudh Please check.

**Comment by jychoudh** [comment]
Looking into this.

**Comment by jychoudh** [comment]
Since the crash doesn't happen always, I couldn't reproduce this. Can you please try again with 1.34.0 release and let us know if you still see the crash?

**Comment by RomanValov** [comment]
This remains reproducible on my side with 1.34.0 release. Here is a bit of findings about reproduction:

Doesn't depend on:
 - Ubuntu distro used -- focal or jammy
 - Running as root or as a user
 - Java source/target version
 - Running using mvn exec:java or java directy
 
For Java 8 installed from Ubuntu packages and SDK 1.31.0 I was unable to reproduce the issue in a docker container. But it reproduces very often without docker (on host). I was unable to start my example on Java 8 with SDK 1.34.0 though.

 For Java 11+ JREs installed from Ubuntu packages there is subjectively higher reproduction rate with Java 17+ JREs as compared to Java 11 JRE.

**Comment by RomanValov** [comment]
May it make sense to share a coredump? I.e. here is a backtrace I can find from it:

```
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140681865029184) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=140681865029184) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=140681865029184, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007ff30dfc0476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007ff30dfa67f3 in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007ff30cb20b9e in __gnu_cxx::__verbose_terminate_handler () at ../../../../src/libstdc++-v3/libsupc++/vterminate.cc:95
#6  0x00007ff30cb2c20c in __cxxabiv1::__terminate (handler=<optimized out>) at ../../../../src/libstdc++-v3/libsupc++/eh_terminate.cc:48
#7  0x00007ff30cb2c277 in std::terminate () at ../../../../src/libstdc++-v3/libsupc++/eh_terminate.cc:58
#8  0x00007ff2ee81d4d4 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.extension.mas.so
#9  0x00007ff2ee81d4e9 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.extension.mas.so
#10 0x00007ff2ee80cb32 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.extension.mas.so
#11 0x00007ff2ee80cbd1 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.extension.mas.so
#12 0x00007ff2eecd4d5c in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#13 0x00007ff2eedb1f38 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#14 0x00007ff2eedb1ff9 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#15 0x00007ff2eecd4d5c in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#16 0x00007ff2eeddfd40 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#17 0x00007ff2eed04e4e in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#18 0x00007ff2eedb1aa3 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#19 0x00007ff2eed8e4ec in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#20 0x00007ff2eedb22ff in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#21 0x00007ff2eecf248d in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#22 0x00007ff2eedb2700 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#23 0x00007ff2eecf248d in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#24 0x00007ff2eed7b995 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#25 0x00007ff2eecf248d in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#26 0x00007ff2eeda0de5 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#27 0x00007ff2eecf248d in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#28 0x00007ff2eed4df17 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#29 0x00007ff2eed5f8d2 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#30 0x00007ff2eed486a4 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#31 0x00007ff2eed4fe38 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#32 0x00007ff2eed5005b in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#33 0x00007ff2eecd4d5c in ?? () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#34 0x00007ff2eecaefe5 in recognizer_handle_release () from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.core.so
#35 0x00007ff2ef42401a in Java_com_microsoft_cognitiveservices_speech_util_SafeHandle_releaseRecognizerHandle ()
   from /home/reddot/pbsupp/tmp/mas/lib/libMicrosoft.CognitiveServices.Speech.java.bindings.so
```

It should give more information with debug symbols for so libraries from SDK.

**Comment by RomanValov** [comment]
Here is an exact code + pom + dockerfile I use to repdoduce:

[mas.issue-2125.tar.gz](https://github.com/Azure-Samples/cognitive-services-speech-sdk/files/13722202/mas.issue-2125.tar.gz)

```
docker build -t speech-demo --build-arg UBUNTU_VERSION=jammy --build-arg JAVA_VERSION=11 .
docker run --init --privileged -it --rm --env-file ./env.file speech-demo
```

**Comment by jychoudh** [comment]
Thanks for sharing the code and dump! I will investigate this further.

**Comment by RomanValov** [comment]
Previously posted coredump backtrace is related to 1.31.0. Here is one from 1.34.0:

```
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=139647947114048) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=139647947114048) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=139647947114048, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007f0253aca476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007f0253ab07f3 in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007f025262ab9e in __gnu_cxx::__verbose_terminate_handler () at ../../../../src/libstdc++-v3/libsupc++/vterminate.cc:95
#6  0x00007f025263620c in __cxxabiv1::__terminate (handler=<optimized out>) at ../../../../src/libstdc++-v3/libsupc++/eh_terminate.cc:48
#7  0x00007f0252636277 in std::terminate () at ../../../../src/libstdc++-v3/libsupc++/eh_terminate.cc:58
#8  0x00007f01ef82f8d4 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.extension.mas.so
#9  0x00007f01ef82f8e9 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.extension.mas.so
#10 0x00007f01ef81c856 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.extension.mas.so
#11 0x00007f01ef81c8f5 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.extension.mas.so
#12 0x00007f02248d566c in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#13 0x00007f02249b52fa in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#14 0x00007f02249b53bb in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#15 0x00007f02248d566c in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#16 0x00007f02249ef854 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#17 0x00007f022490635c in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#18 0x00007f0224999465 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#19 0x00007f022499a11a in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#20 0x00007f02249b56c1 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#21 0x00007f02248f284d in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#22 0x00007f02249b5ac2 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#23 0x00007f02248f284d in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#24 0x00007f022497f143 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#25 0x00007f02248f284d in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#26 0x00007f022498a671 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#27 0x00007f02248f284d in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#28 0x00007f022494085a in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#29 0x00007f0224940921 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#30 0x00007f022494227d in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#31 0x00007f0224942565 in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#32 0x00007f02248d566c in ?? () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#33 0x00007f022489966f in recognizer_handle_release () from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.core.so
#34 0x00007f0225024d7e in Java_com_microsoft_cognitiveservices_speech_util_SafeHandle_releaseRecognizerHandle ()
   from /home/reddot/pbsupp/tmp/mas/lib/1.34.0/libMicrosoft.CognitiveServices.Speech.java.bindings.so
```

Do you need the full coredump file?

**Comment by jychoudh** [comment]
Thanks! I was able to reproduce the issue and figure out the root cause. I will fix it soon.

**Comment by pankopon** [comment]
Internal work item ref. 6806726. Fix will be included in the Speech SDK 1.37.0 release.

**Comment by pankopon** [comment]
The fix has been released (Speech SDK 1.37.0).
